File: php-api/src/Controllers/PortalHomeController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Services\PortalHomeService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/**
 * Ported from Controllers/PortalHomeController.cs. The end-user-facing side of Organisational
 * Portals — gated by RequireAuthMiddleware only (no ProjectMember/OrgAdmin middleware, see
 * routes.php), same shape as WhiteboardController/ChatController/ToDoController: a Portal must be
 * reachable by an org user who belongs to zero projects. See PortalHomeService's own doc comment for
 * the access-check guarantee every action here relies on.
 */
final class PortalHomeController extends BaseController
{
    private function service(): PortalHomeService
    {
        return new PortalHomeService(Database::connection());
    }

    public function listAccessible(Request $request, Response $response): Response
    {
        return $this->json($response, $this->service()->listAccessible($this->callerOrgId($request), $this->callerUserId($request)));
    }

    public function getBySlug(Request $request, Response $response, array $args): Response
    {
        $portal = $this->service()->getBySlug($this->callerOrgId($request), $args['slug'], $this->callerUserId($request));
        return $portal === null ? $this->notFound($response) : $this->json($response, $portal);
    }

    public function listAvailableForms(Request $request, Response $response, array $args): Response
    {
        $forms = $this->service()->listAvailableForms($this->callerOrgId($request), $args['portalId'], $this->callerUserId($request));
        return $forms === null ? $this->notFound($response) : $this->json($response, $forms);
    }

    public function listMySubmissions(Request $request, Response $response, array $args): Response
    {
        $submissions = $this->service()->listMySubmissions($this->callerOrgId($request), $args['portalId'], $this->callerUserId($request));
        return $submissions === null ? $this->notFound($response) : $this->json($response, $submissions);
    }

    public function listQa(Request $request, Response $response, array $args): Response
    {
        $qa = $this->service()->listQa($this->callerOrgId($request), $args['portalId'], $this->callerUserId($request));
        return $qa === null ? $this->notFound($response) : $this->json($response, $qa);
    }

    public function createSubmission(Request $request, Response $response, array $args): Response
    {
        $body = $this->body($request);
        $submission = $this->service()->createSubmission($this->callerOrgId($request), $args['portalId'], $this->callerUserId($request), $body);
        return $submission === null ? $this->notFound($response) : $this->json($response, $submission);
    }

    public function updateSubmission(Request $request, Response $response, array $args): Response
    {
        $body = $this->body($request);
        $submission = $this->service()->updateSubmission($this->callerOrgId($request), $args['portalId'], $this->callerUserId($request), $args['submissionId'], $body);
        return $submission === null ? $this->notFound($response) : $this->json($response, $submission);
    }

    public function deleteSubmission(Request $request, Response $response, array $args): Response
    {
        $deleted = $this->service()->deleteSubmission($this->callerOrgId($request), $args['portalId'], $this->callerUserId($request), $args['submissionId']);
        return $deleted ? $this->noContent($response) : $this->notFound($response);
    }

    public function submitSubmission(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->submitSubmission($this->callerOrgId($request), $args['portalId'], $this->callerUserId($request), $args['submissionId']);
        if (!$result['ok']) {
            return $result['error'] === 'not_found' ? $this->notFound($response) : $this->json($response, ['message' => $result['error']], 400);
        }
        return $this->json($response, $result['dto']);
    }
}

File: api/Enkl.Api/Controllers/PortalHomeController.cs
using Enkl.Api.Auth;
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Enkl.Api.Controllers;

/// <summary>
/// The end-user-facing side of Organisational Portals — [Authorize] only at class level (no
/// ProjectMember/OrgAdmin policy), same shape as WhiteboardController/ChatController/ToDoController:
/// a Portal must be reachable by an org user who belongs to zero projects. See PortalHomeService's
/// own doc comment for the access-check guarantee every action here relies on.
/// </summary>
[ApiController]
[Authorize]
[Route("api/portals")]
public class PortalHomeController : ControllerBase
{
    private readonly PortalHomeService _home;

    public PortalHomeController(PortalHomeService home)
    {
        _home = home;
    }

    // Backs the side nav's "Portals" section — every published Portal in the caller's own org this
    // user actually has access to. GET, not the same route shape as GetBySlug below (no {slug}
    // segment), so the two never collide.
    [HttpGet]
    public async Task<IActionResult> ListAccessible()
    {
        return Ok(await _home.ListAccessibleAsync(User.OrgId(), User.UserId()));
    }

    [HttpGet("{slug}")]
    public async Task<IActionResult> GetBySlug(string slug)
    {
        var portal = await _home.GetBySlugAsync(User.OrgId(), slug, User.UserId());
        return portal is null ? NotFound() : Ok(portal);
    }

    [HttpGet("{portalId:guid}/forms")]
    public async Task<IActionResult> ListAvailableForms(Guid portalId)
    {
        var forms = await _home.ListAvailableFormsAsync(User.OrgId(), portalId, User.UserId());
        return forms is null ? NotFound() : Ok(forms);
    }

    [HttpGet("{portalId:guid}/submissions")]
    public async Task<IActionResult> ListMySubmissions(Guid portalId)
    {
        var submissions = await _home.ListMySubmissionsAsync(User.OrgId(), portalId, User.UserId());
        return submissions is null ? NotFound() : Ok(submissions);
    }

    [HttpGet("{portalId:guid}/qa")]
    public async Task<IActionResult> ListQa(Guid portalId)
    {
        var qa = await _home.ListQaAsync(User.OrgId(), portalId, User.UserId());
        return qa is null ? NotFound() : Ok(qa);
    }

    [HttpGet("{portalId:guid}/submissions/{submissionId:guid}")]
    public async Task<IActionResult> GetSubmission(Guid portalId, Guid submissionId)
    {
        var submission = await _home.GetSubmissionAsync(User.OrgId(), portalId, User.UserId(), submissionId);
        return submission is null ? NotFound() : Ok(submission);
    }

    [HttpPost("{portalId:guid}/submissions")]
    public async Task<IActionResult> CreateSubmission(Guid portalId, CreateFormSubmissionRequest request)
    {
        var submission = await _home.CreateSubmissionAsync(User.OrgId(), portalId, User.UserId(), request);
        return submission is null ? NotFound() : Ok(submission);
    }

    [HttpPut("{portalId:guid}/submissions/{submissionId:guid}")]
    public async Task<IActionResult> UpdateSubmission(Guid portalId, Guid submissionId, UpdateFormSubmissionRequest request)
    {
        var submission = await _home.UpdateSubmissionAsync(User.OrgId(), portalId, User.UserId(), submissionId, request);
        return submission is null ? NotFound() : Ok(submission);
    }

    [HttpDelete("{portalId:guid}/submissions/{submissionId:guid}")]
    public async Task<IActionResult> DeleteSubmission(Guid portalId, Guid submissionId)
    {
        var deleted = await _home.DeleteSubmissionAsync(User.OrgId(), portalId, User.UserId(), submissionId);
        return deleted ? NoContent() : NotFound();
    }

    [HttpPost("{portalId:guid}/submissions/{submissionId:guid}/submit")]
    public async Task<IActionResult> SubmitSubmission(Guid portalId, Guid submissionId)
    {
        var (ok, error, dto) = await _home.SubmitSubmissionAsync(User.OrgId(), portalId, User.UserId(), submissionId);
        if (!ok) return error == "not_found" ? NotFound() : BadRequest(new { message = error });
        return Ok(dto);
    }
}
